### PR TITLE
Force verification even for refreshed clients

### PR DESCRIPTION
**Pull request by dbkr**
Set a flag on login to remember that the device needs to be verified so that we don't forget if the user refreshes the page, but still allow user with an existing unverified session to stay logged in.

With apologies to the reviewer, this involved changing the flow of what screens are shown when in MatrixChat, which turns out to be somewhat of a jenga tower, so most of the changes here are around getting the tests to pass again, including one to make the post-registration checkboxes not fail if they're displayed before the push rules are fetched. I can split that bit out but it's probably not worth it: the MatrixChat test fixes would be harder.

## Checklist

-   [ ] Tests written for new code (and old code if feasible).
-   [ ] New or updated `public`/`exported` symbols have accurate [TSDoc](https://tsdoc.org/) documentation.
-   [ ] Linter and other CI checks pass.
-   [ ] I have licensed the changes to Element by completing the [Contributor License Agreement (CLA)](https://cla-assistant.io/element-hq/matrix-react-sdk)

### CODE REVIEW

File: src/components/structures/MatrixChat.tsx [hunk 1]
```diff
@@ -881,7 +892,7 @@ export default class MatrixChat extends React.PureComponent<IProps, IState> {
                 });
                 break;
             case "client_started":
-                this.onClientStarted();
+                this.onClientStarted().then();
```
Comment (richvdh) [inline src/components/structures/MatrixChat.tsx:895]: driveby comment: what's with the empty `then()` ?

File: src/hooks/useUserOnboardingContext.ts [hunk 2]
```diff
@@ -77,14 +77,37 @@ function useUserOnboardingContextValue<T>(defaultValue: T, callback: (cli: Matri
...
     const setting = useSettingValue("notificationsEnabled");
     useEffect(() => {
-        setValue(Notifier.shouldShowPrompt());
-    }, [setting]);
+        updateValue();
+    }, [setting, updateValue]);
+
+    // shouldShowPrompt is dependent on the client having push rules. There isn't an event for the client
+    // fetching its push rules, but we'll know it has them by the time it sync, so we update this on sync.
+    useEffect(() => {
```
Comment (florianduros) [inline src/hooks/useUserOnboardingContext.ts:99]: Can we use `useTypedEventEmitter` or `useEventEmitter`?

File: test/components/structures/MatrixChat-test.tsx [hunk 3]
```diff
@@ -358,6 +386,8 @@ describe("<MatrixChat />", () => {
             await flushPromises();
 
             expect(completeAuthorizationCodeGrant).toHaveBeenCalledWith(code, state);
+
+            //await waitFor(() => expect(screen.queryByTestId("spinner")).not.toBeInTheDocument());
```
Comment (florianduros) [inline test/components/structures/MatrixChat-test.tsx:390]: Remaining commented code

**Review by florianduros — CHANGES_REQUESTED** [review]

File: src/components/structures/MatrixChat.tsx (hunk 1 above)
Comment (dbkr) [reply src/components/structures/MatrixChat.tsx:895]: Just as a way of indicating that we're deliberately not waiting for an async function, but I guess it didn't work. :) I forget if we decided on a preferred way of doing this - `(void)this.onClientStarted();` maybe? Or just a comment?

File: package.json [hunk 4]
```diff
@@ -198,6 +198,7 @@
         "axe-core": "4.10.0",
         "babel-jest": "^29.0.0",
         "blob-polyfill": "^9.0.0",
+        "core-js": "^3.38.1",
```
[the remark this answers is not in the document]
Comment (dbkr) [reply package.json:201]: This is to add a polyfill to avoid the test crashing, ie. https://github.com/element-hq/matrix-react-sdk/pull/44/files#diff-0bd42c32df0a1d0f3ebbfb8eebc1c08958e613a9e85ff581f99ccd7f00b4ea34R11

File: playwright/e2e/login/login.spec.ts [hunk 5]
```diff
@@ -65,17 +130,97 @@ test.describe("Login", () => {
...
+                homeserver,
+                request,
+            }) => {
+                const res = await request.post(
+                    `${homeserver.config.baseUrl}/_matrix/client/v3/keys/device_signing/upload`,
+                    { headers: { Authorization: `Bearer ${creds.accessToken}` }, data: DEVICE_SIGNING_KEYS_BODY },
+                );
+                if (res.status() / 100 !== 2) {
+                    console.log(await res.json());
+                }
```
[the remark this answers is not in the document]
Comment (dbkr) [reply playwright/e2e/login/login.spec.ts:151]: Actually these were intentional, but I can add a log line so it's more obvious.

File: playwright/e2e/login/login.spec.ts [hunk 6]
```diff
@@ -65,17 +130,97 @@ test.describe("Login", () => {
...
+                    homeserver,
+                    request,
+                }) => {
+                    const res = await request.post(
+                        `${homeserver.config.baseUrl}/_matrix/client/v3/keys/device_signing/upload`,
+                        { headers: { Authorization: `Bearer ${creds.accessToken}` }, data: DEVICE_SIGNING_KEYS_BODY },
+                    );
+                    if (res.status() / 100 !== 2) {
+                        console.log(await res.json());
+                    }
```
[the remark this answers is not in the document]
Comment (dbkr) [reply playwright/e2e/login/login.spec.ts:180]: As above

File: playwright/e2e/login/login.spec.ts [hunk 7]
```diff
@@ -65,17 +130,97 @@ test.describe("Login", () => {
...
+                    request,
+                }) => {
+                    console.log(`uid ${creds.userId} body`, DEVICE_SIGNING_KEYS_BODY);
+                    const res = await request.post(
+                        `${homeserver.config.baseUrl}/_matrix/client/v3/keys/device_signing/upload`,
+                        { headers: { Authorization: `Bearer ${creds.accessToken}` }, data: DEVICE_SIGNING_KEYS_BODY },
+                    );
+                    if (res.status() / 100 !== 2) {
+                        console.log(await res.json());
+                    }
```
[the remark this answers is not in the document]
Comment (dbkr) [reply playwright/e2e/login/login.spec.ts:211]: As above

File: src/components/structures/MatrixChat.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/structures/MatrixChat.tsx:895]: rich commented the same: I've responded there: https://github.com/element-hq/matrix-react-sdk/pull/44#discussion_r1782413242

Comment (florianduros) [reply src/components/structures/MatrixChat.tsx:895]: A comment to explain why it's needed is welcomed :)

File: src/components/structures/MatrixChat.tsx [hunk 8]
```diff
@@ -1328,30 +1353,7 @@ export default class MatrixChat extends React.PureComponent<IProps, IState> {
...
-                (originalSettingName, changedInRoomId, atLevel, newValueAtLevel, newValue) => {
-                    if (newValue !== null && this.state.view === Views.USE_CASE_SELECTION) {
-                        this.onShowPostLoginScreen();
-                    }
-                },
-            );
-        } else {
-            return this.onShowPostLoginScreen();
-        }
+        this.onShowPostLoginScreen().then();
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/structures/MatrixChat.tsx:1356]: As above

**Review by florianduros — APPROVED** [review]
